File: Python/SmallestStack.py
#-*- coding: utf-8 -*-

queue = []
minIndex = []

def EeQueue(queueValue):
	queue.append(queueValue)
	if len(minIndex) == 0:
		minIndex.insert(0, len(queue) - 1)
	else:
		i = len(minIndex)-1;
		while (i >= 0):
			if i > 0 and queue[minIndex[i]] < queueValue and queue[minIndex[i - 1]] >= queueValue:
				minIndex.insert(i - 1, len(queue) - 1)
				break
			elif queue[minIndex[i]] >= queueValue:
				minIndex.insert(i + 1, len(queue) - 1)
				break
			elif i == 0:
				minIndex.insert(0, len(queue) - 1)
				break
			i = i - 1

def DeQueue(queueValue):
	if len(queue) <= 0:
		return
	queueIndex = queue.index(queueValue);
	if queueIndex >= 0:
		queue.remove(queueValue)
		minIndex.remove(queueIndex)
		delCount = len(queue) - queueIndex
		i = len(minIndex) - 1;
		while delCount > 0 and i >= 0:
			if minIndex[i] > queueIndex:
				minIndex[i] = minIndex[i] - 1
				delCount -= 1
			i = i - 1

def Min():
	if len(minIndex) <= 0:
		return
	index = minIndex[len(minIndex) - 1]
	minValue = queue[index]
	DeQueue(minValue)
	return minValue;

def Max():
	if len(minIndex) <= 0:
		return
	index = minIndex[0]
	maxValue = queue[index]
	DeQueue(maxValue)
	return maxValue;


EeQueue(2);
print (queue);
EeQueue(3);
EeQueue(1);
EeQueue(5);
print (queue);
print (Min());
DeQueue(3);
print (Min());
print (Min());
print (Min());
print (Min());

File: C#/SmallestStack.cs
public class SmallestStack 
{
	private List<int> minIndex = new List<int>();
	private List<float> queue = new List<float>();

	public void EeQueue(float queueValue)
	{
		queue.add(queueValue);
		if (minIndex.count == 0)
		{
			minIndex.Insert(0, queue.Count - 1);
		}
		else
		{
			int i = minIndex.Count-1;
			for (; i >= 0; i--) 
			{
				if (i > 0 && queue[minIndex[i]] < queueValue && queue[minIndex[i - 1]] >= queueValue)
				{
					minIndex.Insert(i - 1, queue.Count - 1);
					break;
				}
				else if (queue[minIndex[i]] >= queueValue)
				{
					minIndex.Insert(i + 1, queue.Count - 1);
					break;
				}
				else if (i == 0)
				{
					minIndex.Insert(0, queue.Count - 1);
					break;
				}

			}
		}
	}

	public void DeQueue(float queueValue)
	{
		if (queue.Count <= 0) return;
		int index = queue.findIndex(queueValue);
		if (index >= 0)
		{
			queue.Remove(queueValue);
			minIndex.Remove(index);
			int i = minIndex.length - 1;
			for (i; i >= 0; i--)
			{
				if (minIndex[i] > index)
				{
					minIndex[i] = minIndex[i] - 1;
				}
			}
		}
	}

	public float Min
	{
		get{
			if (queue.Count <= 0) return null;
			int index = minIndex[minIndex.length - 1];
			float minValue = queue[index];
			DeQueue(minValue);
			return minValue;
		}
	}

	public float Max
	{
		get{
			if (queue.Count <= 0) return null;
			int index = minIndex[0];
			float maxValue = queue[index];
			DeQueue(maxValue);
			return maxValue;
		}
	}
}